tower-util/src/combinators/map_err.rs

```diff
@@ -0,0 +1,72 @@
+use futures_util::{future::MapErr as MapErrFut, TryFutureExt};
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+#[derive(Debug)]
+pub struct MapErr<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapErr<S, F> {
+    pub fn new(inner: S, f: F) -> Self {
+        MapErr { f, inner }
+    }
+}
+
+impl<S, F, Request, Error> Service<Request> for MapErr<S, F>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error,
+    F: Clone,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = MapErrFut<S::Future, F>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(self.f.clone())
+    }
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        self.inner.call(request).map_err(self.f.clone())
+    }
+}
+
+#[derive(Debug)]
+pub struct MapErrLayer<F, Request, Error> {
+    f: F,
+    request: PhantomData<Request>,
+    error: PhantomData<Error>,
+}
+
+impl<F, Request, Error> MapErrLayer<F, Request, Error>
+where
+    F: Clone,
+{
+    pub fn new(f: F) -> Self {
+        MapErrLayer {
+            f,
+            request: PhantomData::default(),
+            error: PhantomData::default(),
```

Comment:
nit: this can just be
```suggestion
            request: PhantomData,
            error: PhantomData,
```